KeyError when using numpy

Hi,

Ive been trying to optimize a breadth-first-search type algorithm by compiling it with Pythran, and I get an error message, first I get the warning

```
/home/vincent/.local/lib/python3.5/site-packages/pythran/interval.py:107: RuntimeWarning: invalid value encountered in double_scalars
  res = [v1 // v2 for v1, v2 in

```
then eventually

`KeyError: <gast.gast.Tuple object at 0x7f388ed31f60>`
 

I was using the queue class in the Python but Ive re-written this as a numpy array, so as far as I can tell this is all just numpy and python - although it is Python 3.
Ive tested running it with just Python since I re-wrote it, which it appears to do fine.  

The code is as follows:
```
import numpy as np

def addToQueue(q, end, size, x, y, z):
    q[end, 0] = x
    q[end, 1] = y
    q[end, 2] = z
    return (end + 1) % size

def removeFromQueue(q, start, size):
    return q[start], (start + 1) % size

#pythran export updateGeodesicsOpt(float, float, float, float[][][], float[][][][], float[][][], float[][][][])
def updateGeodesicsOpt(maxPathLength, stdDev, minEdge, img, probs, segmentation, weighting):
    qStart = np.array([0,0])
    qEnd   = np.array([0,0])
    maxSize = img.shape[0]*img.shape[1]*img.shape[2]
    qs = np.zeros([2] + [maxSize] + [3], dtype=np.int8)
    
    for i, row in enumerate(probs[0]):
        for j, col in enumerate(row):
            for k, val in enumerate(col):
                if (val < 0):
                    index = int(round(val))+2
                    qEnd[index] = addToQueue(qs[index], qEnd[index], maxSize, i, j, k)

    directions = np.array([ [1,0,0], [-1,0,0], [0,1,0], [0,-1,0], [0,0,1], [0,0,-1] ])
    for index in [0,1]:
        pathLengths = np.zeros(img.shape)#fill(img.shape, fill_value=-1.0)
        computedPathLengths = np.zeros(img.shape)#np.fill(img.shape, fill_value=-1.0)
        for i, row in enumerate(img):
            for j, col in enumerate(row):
                for k, val in enumerate(col):
                    pathLengths[i,j,k] = -1.0
                    computedPathLengths[i,j,k] = -1.0
        q = qs[index]
        while (qStart[index] != qEnd[index]):
            coords, qStart[index] = removeFromQueue(q, qStart[index], maxSize)

            row, col, depth = coords
            currPathLength = pathLengths[row, col, depth]
            # The if statement below is mean to skip over things in the queue that we have already computed
            if (computedPathLengths[row, col, depth] >= 0 and
                abs(currPathLength - computedPathLengths[row, col, depth]) < 0.0001):
                # If weve got a path length saved and it is the same as the current path length
                continue
            else:
                # Save path length
                if (currPathLength < 0):
                    currPathLength = 0
                computedPathLengths[row,col,depth] = currPathLength
            for direction in directions:
                # Go in every direction
                newRow = row+direction[0]
                newCol = col+direction[1]
                newDepth = depth+direction[2]
                if (newRow < 0 or newCol < 0 or newDepth < 0):
                    # Dont go to negative indices
                    continue
                elif (newRow >= img.shape[0] or newCol >= img.shape[1]
                      or newDepth >= img.shape[2]):
                    continue
                otherVal = segmentation[newRow, newCol, newDepth]
                if (probs[0, newRow, newCol, newDepth] >= 0):# and index != otherVal:
                    # If this is a normal (non-scrible) and it has the opposite marking - does it need to have the opposite marking? Defeats the purpose maybe
                    i1 = img[row, col, depth]
                    i2 = img[newRow, newCol, newDepth]
                    length = 1 - np.exp(-(i2 - i1)**2/(2*stdDev**2)) + minEdge
                    newPathLength = currPathLength + length
                    if (newPathLength > maxPathLength):
                            # If weve gone over the maximum path length, abandon
                        continue
                    destPathLength = pathLengths[newRow, newCol, newDepth]
                    if (newPathLength < destPathLength*0.8
                        or newPathLength < destPathLength - 0.05
                        or destPathLength < 0):
                        # If weve made a shorter path, save it and place in queue
                        qEnd[index] = addToQueue(q, qEnd[index], maxSize, newRow, newCol, newDepth)
                        pathLengths[newRow, newCol, newDepth] = newPathLength
                        if index != otherVal:
                            weighting[0, newRow, newCol, newDepth] = 0
                            weighting[1, newRow, newCol, newDepth] = 0
```

and the full error message:

```
`/home/vincent/.local/lib/python3.5/site-packages/pythran/interval.py:107: RuntimeWarning: invalid value encountered in double_scalars
  res = [v1 // v2 for v1, v2 in
Traceback (most recent call last):
  File "/home/vincent/.local/bin/pythran", line 11, in <module>
    sys.exit(run())
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/run.py", line 152, in run
    **compile_flags(args))
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/toolchain.py", line 479, in compile_pythranfile
    **kwargs)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/toolchain.py", line 411, in compile_pythrancode
    module_dir)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/toolchain.py", line 172, in generate_cxx
    optimizations, module_dir)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/toolchain.py", line 140, in front_middle_end
    refine(pm, ir, optimizations)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/middlend.py", line 49, in refine
    apply_optimisation |= pm.apply(optimization, node)[0]
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/passmanager.py", line 217, in apply
    res = a.apply(node, ctx)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/passmanager.py", line 178, in apply
    new_node = self.run(node, ctx)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/passmanager.py", line 171, in run
    n = super(Transformation, self).run(node, ctx)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/passmanager.py", line 113, in run
    self.prepare(node, ctx)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/passmanager.py", line 104, in prepare
    result = d.run(rnode, ctx)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/analyses/pure_expressions.py", line 71, in run
    super(PureExpressions, self).prepare(node, ctx)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/passmanager.py", line 104, in prepare
    result = d.run(rnode, ctx)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/analyses/argument_effects.py", line 75, in run
    super(ArgumentEffects, self).run(node, ctx)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/passmanager.py", line 130, in run
    super(Analysis, self).run(node, ctx)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/passmanager.py", line 114, in run
    return self.visit(node)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/passmanager.py", line 80, in visit
    return super(ContextManager, self).visit(node)
  File "/usr/lib/python3.5/ast.py", line 245, in visit
    return visitor(node)
  File "/usr/lib/python3.5/ast.py", line 253, in generic_visit
    self.visit(item)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/passmanager.py", line 80, in visit
    return super(ContextManager, self).visit(node)
  File "/usr/lib/python3.5/ast.py", line 245, in visit
    return visitor(node)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/analyses/argument_effects.py", line 115, in visit_FunctionDef
    self.generic_visit(node)
  File "/usr/lib/python3.5/ast.py", line 253, in generic_visit
    self.visit(item)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/passmanager.py", line 80, in visit
    return super(ContextManager, self).visit(node)
  File "/usr/lib/python3.5/ast.py", line 245, in visit
    return visitor(node)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/analyses/argument_effects.py", line 135, in visit_Assign
    self.generic_visit(node)
  File "/usr/lib/python3.5/ast.py", line 255, in generic_visit
    self.visit(value)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/passmanager.py", line 80, in visit
    return super(ContextManager, self).visit(node)
  File "/usr/lib/python3.5/ast.py", line 245, in visit
    return visitor(node)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/analyses/argument_effects.py", line 139, in visit_Call
    n = self.argument_index(arg)
  File "/home/vincent/.local/lib/python3.5/site-packages/pythran/analyses/argument_effects.py", line 101, in argument_index
    for node_alias in self.aliases[node]:
KeyError: <gast.gast.Tuple object at 0x7f388ed31f60>

```

Any help would be very much appreciated
Thanks

Thanks for the help! Brilliant, that has seem to fix the original error - did you manage to get it to compile though? 

Was the line meant to be
`qs = np.zeros((2, maxSize, 3), dtype=int)`
or
`qs = np.zeros((2, maxSize, 3), dtype=np.int8)`

I now get a bunch of C compilation error depending on which I use. If I use int, I get 

```
/home/vincent/.local/lib/python3.5/site-packages/pythran/interval.py:107: RuntimeWarning: invalid value encountered in double_scalars
  res = [v1 // v2 for v1, v2 in
/tmp/tmptz35umgp.cpp: In instantiation of ‘typename __pythran_geodesicsOpt::updateGeodesicsOpt::type<argument_type0, argument_type1, argument_type2, argument_type3, argument_type4, argument_type5, argument_type6>::result_type __pythran_geodesicsOpt::updateGeodesicsOpt::operator()(argument_type0&&, argument_type1&&, argument_type2&&, argument_type3&&, argument_type4&&, argument_type5&&, argument_type6&&) const [with argument_type0 = double&; argument_type1 = double&; argument_type2 = double&; argument_type3 = {anonymous}::pythonic::types::ndarray<double, 3ul>&; argument_type4 = {anonymous}::pythonic::types::ndarray<double, 4ul>&; argument_type5 = {anonymous}::pythonic::types::ndarray<double, 3ul>&; argument_type6 = {anonymous}::pythonic::types::ndarray<double, 4ul>&; typename __pythran_geodesicsOpt::updateGeodesicsOpt::type<argument_type0, argument_type1, argument_type2, argument_type3, argument_type4, argument_type5, argument_type6>::result_type = {anonymous}::pythonic::types::none_type]’:
/tmp/tmptz35umgp.cpp:508:156:   required from here
/tmp/tmptz35umgp.cpp:401:7: error: invalid initialization of non-const reference of type ‘long int&’ from an rvalue of type ‘long int’
       for (auto&& index: pythonic::__builtin__::functor::xrange{}(0L, 2L, 1L))
       ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/include/types/lazy.hpp:17: confused by earlier errors, bailing out
/tmp/tmptz35umgp.cpp: In instantiation of ‘typename __pythran_geodesicsOpt::updateGeodesicsOpt::type<argument_type0, argument_type1, argument_type2, argument_type3, argument_type4, argument_type5, argument_type6>::result_type __pythran_geodesicsOpt::updateGeodesicsOpt::operator()(argument_type0&&, argument_type1&&, argument_type2&&, argument_type3&&, argument_type4&&, argument_type5&&, argument_type6&&) const [with argument_type0 = double&; argument_type1 = double&; argument_type2 = double&; argument_type3 = {anonymous}::pythonic::types::ndarray<double, 3ul>&; argument_type4 = {anonymous}::pythonic::types::ndarray<double, 4ul>&; argument_type5 = {anonymous}::pythonic::types::ndarray<double, 3ul>&; argument_type6 = {anonymous}::pythonic::types::ndarray<double, 4ul>&; typename __pythran_geodesicsOpt::updateGeodesicsOpt::type<argument_type0, argument_type1, argument_type2, argument_type3, argument_type4, argument_type5, argument_type6>::result_type = {anonymous}::pythonic::types::none_type]’:
/tmp/tmptz35umgp.cpp:508:156:   required from here
/tmp/tmptz35umgp.cpp:401:7: error: invalid initialization of non-const reference of type ‘long int&’ from an rvalue of type ‘long int’
       for (auto&& index: pythonic::__builtin__::functor::xrange{}(0L, 2L, 1L))
       ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/include/types/lazy.hpp:17: confused by earlier errors, bailing out
WARNING: Compilation error, trying hard to find its origin...
WARNING: Nop, I'm going to flood you with C++ errors!
CRITICAL: Cover me Jack. Jack? Jaaaaack!!!!
E: error: Command "x86_64-linux-gnu-g++ -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DUSE_GMP -DENABLE_PYTHON_MODULE -D__PYTHRAN__=3 -DPYTHRAN_BLAS_BLAS -I/home/vincent/.local/lib/python3.5/site-packages/pythran -I/home/vincent/.local/lib/python3.5/site-packages/numpy/core/include -I/usr/local/include -I/usr/include -I/home/vincent/.local/lib/python3.5/site-packages/numpy/core/include -I/usr/include/python3.5m -c /tmp/tmptz35umgp.cpp -o /tmp/tmpd48bvhdf/tmp/tmptz35umgp.o -std=c++11 -fno-math-errno -w -fvisibility=hidden -fno-wrapv" failed with exit status 1

```

and with np.int8

```
/home/vincent/.local/lib/python3.5/site-packages/pythran/interval.py:107: RuntimeWarning: invalid value encountered in double_scalars
  res = [v1 // v2 for v1, v2 in
In file included from /tmp/tmptvimr55i.cpp:12:0:
/tmp/tmptvimr55i.cpp: In instantiation of ‘typename __pythran_geodesicsOpt::updateGeodesicsOpt::type<argument_type0, argument_type1, argument_type2, argument_type3, argument_type4, argument_type5, argument_type6>::result_type __pythran_geodesicsOpt::updateGeodesicsOpt::operator()(argument_type0&&, argument_type1&&, argument_type2&&, argument_type3&&, argument_type4&&, argument_type5&&, argument_type6&&) const [with argument_type0 = double&; argument_type1 = double&; argument_type2 = double&; argument_type3 = {anonymous}::pythonic::types::ndarray<double, 3ul>&; argument_type4 = {anonymous}::pythonic::types::ndarray<double, 4ul>&; argument_type5 = {anonymous}::pythonic::types::ndarray<double, 3ul>&; argument_type6 = {anonymous}::pythonic::types::ndarray<double, 4ul>&; typename __pythran_geodesicsOpt::updateGeodesicsOpt::type<argument_type0, argument_type1, argument_type2, argument_type3, argument_type4, argument_type5, argument_type6>::result_type = {anonymous}::pythonic::types::none_type]’:
/tmp/tmptvimr55i.cpp:512:156:   required from here
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/tmp/tmptvimr55i.cpp:405:7: error: invalid initialization of non-const reference of type ‘long int&’ from an rvalue of type ‘long int’
       for (auto&& index_: pythonic::__builtin__::functor::xrange{}(0L, 2L, 1L))
       ^
In file included from /tmp/tmptvimr55i.cpp:12:0:
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/include/types/lazy.hpp:17: confused by earlier errors, bailing out
In file included from /tmp/tmptvimr55i.cpp:12:0:
/tmp/tmptvimr55i.cpp: In instantiation of ‘typename __pythran_geodesicsOpt::updateGeodesicsOpt::type<argument_type0, argument_type1, argument_type2, argument_type3, argument_type4, argument_type5, argument_type6>::result_type __pythran_geodesicsOpt::updateGeodesicsOpt::operator()(argument_type0&&, argument_type1&&, argument_type2&&, argument_type3&&, argument_type4&&, argument_type5&&, argument_type6&&) const [with argument_type0 = double&; argument_type1 = double&; argument_type2 = double&; argument_type3 = {anonymous}::pythonic::types::ndarray<double, 3ul>&; argument_type4 = {anonymous}::pythonic::types::ndarray<double, 4ul>&; argument_type5 = {anonymous}::pythonic::types::ndarray<double, 3ul>&; argument_type6 = {anonymous}::pythonic::types::ndarray<double, 4ul>&; typename __pythran_geodesicsOpt::updateGeodesicsOpt::type<argument_type0, argument_type1, argument_type2, argument_type3, argument_type4, argument_type5, argument_type6>::result_type = {anonymous}::pythonic::types::none_type]’:
/tmp/tmptvimr55i.cpp:512:156:   required from here
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/tmp/tmptvimr55i.cpp:405:7: error: invalid initialization of non-const reference of type ‘long int&’ from an rvalue of type ‘long int’
       for (auto&& index_: pythonic::__builtin__::functor::xrange{}(0L, 2L, 1L))
       ^
In file included from /tmp/tmptvimr55i.cpp:12:0:
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:695:64: note: candidate 1: typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type {anonymous}::pythonic::types::ndarray<T, N>::operator[](const F&) const [with F = {anonymous}::pythonic::types::array<signed char, 3ul>; T = double; long unsigned int N = 3ul; typename std::enable_if<(({anonymous}::pythonic::types::is_numexpr_arg<F>::value && (! {anonymous}::pythonic::types::is_array_index<F>::value)) && (! std::is_same<bool, typename F::dtype>::value)), {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<T, N>, F> >::type = {anonymous}::pythonic::types::numpy_vexpr<{anonymous}::pythonic::types::ndarray<double, 3ul>, {anonymous}::pythonic::types::array<signed char, 3ul> >]
                           numpy_vexpr<ndarray<T, N>, F>>::type ndarray<T, N>::
                                                                ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/types/ndarray.hpp:519:6: note: candidate 2: T& {anonymous}::pythonic::types::ndarray<T, N>::operator[](const {anonymous}::pythonic::types::array<long int, Value>&) [with T = double; long unsigned int N = 3ul]
   T &ndarray<T, N>::operator[](array<long, N> const &indices)
      ^
/home/vincent/.local/lib/python3.5/site-packages/pythran/pythonic/include/types/lazy.hpp:17: confused by earlier errors, bailing out
WARNING: Compilation error, trying hard to find its origin...
WARNING: Nop, I'm going to flood you with C++ errors!
CRITICAL: Cover me Jack. Jack? Jaaaaack!!!!
E: error: Command "x86_64-linux-gnu-g++ -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DUSE_GMP -DENABLE_PYTHON_MODULE -D__PYTHRAN__=3 -DPYTHRAN_BLAS_BLAS -I/home/vincent/.local/lib/python3.5/site-packages/pythran -I/home/vincent/.local/lib/python3.5/site-packages/numpy/core/include -I/usr/local/include -I/usr/include -I/home/vincent/.local/lib/python3.5/site-packages/numpy/core/include -I/usr/include/python3.5m -c /tmp/tmptvimr55i.cpp -o /tmp/tmpt7h0_msj/tmp/tmptvimr55i.o -std=c++11 -fno-math-errno -w -fvisibility=hidden -fno-wrapv" failed with exit status 1
```

--- a/pythran/pythonic/include/types/ndarray.hpp
+++ b/pythran/pythonic/include/types/ndarray.hpp
@@ -421,21 +421,23 @@ namespace types
     template <class F> // indexing through an array of boolean -- a mask
     typename std::enable_if<
         is_numexpr_arg<F>::value &&
-            std::is_same<bool, typename F::dtype>::value && F::value == 1,
+            std::is_same<bool, typename F::dtype>::value && F::value == 1 &&
+            !is_pod_array<F>::value,
         numpy_vexpr<ndarray, ndarray<long, pshape<long>>>>::type
     fast(F const &filter) const;
 
     template <class F> // indexing through an array of boolean -- a mask
     typename std::enable_if<
         is_numexpr_arg<F>::value &&
-            std::is_same<bool, typename F::dtype>::value && F::value == 1,
+            std::is_same<bool, typename F::dtype>::value && F::value == 1 &&
+            !is_pod_array<F>::value,
         numpy_vexpr<ndarray, ndarray<long, pshape<long>>>>::type
     operator[](F const &filter) const;
 
     template <class F> // indexing through an array of boolean -- a mask
     typename std::enable_if<is_numexpr_arg<F>::value &&
                                 std::is_same<bool, typename F::dtype>::value &&
-                                F::value != 1,
+                                F::value != 1 && !is_pod_array<F>::value,
                             numpy_vexpr<ndarray<T, pshape<long>>,
                                         ndarray<long, pshape<long>>>>::type
     fast(F const &filter) const;
@@ -443,7 +445,7 @@ namespace types
     template <class F> // indexing through an array of boolean -- a mask
     typename std::enable_if<is_numexpr_arg<F>::value &&
                                 std::is_same<bool, typename F::dtype>::value &&
-                                F::value != 1,
+                                F::value != 1 && !is_pod_array<F>::value,
                             numpy_vexpr<ndarray<T, pshape<long>>,
                                         ndarray<long, pshape<long>>>>::type
     operator[](F const &filter) const;
@@ -451,14 +453,16 @@ namespace types
     template <class F> // indexing through an array of indices -- a view
     typename std::enable_if<is_numexpr_arg<F>::value &&
                                 !is_array_index<F>::value &&
-                                !std::is_same<bool, typename F::dtype>::value,
+                                !std::is_same<bool, typename F::dtype>::value &&
+                                !is_pod_array<F>::value,
                             numpy_vexpr<ndarray, F>>::type
     operator[](F const &filter) const;
 
     template <class F> // indexing through an array of indices -- a view
     typename std::enable_if<is_numexpr_arg<F>::value &&
                                 !is_array_index<F>::value &&
-                                !std::is_same<bool, typename F::dtype>::value,
+                                !std::is_same<bool, typename F::dtype>::value &&
+                                !is_pod_array<F>::value,
                             numpy_vexpr<ndarray, F>>::type
     fast(F const &filter) const;
 

--- a/pythran/pythonic/include/types/numpy_expr.hpp
+++ b/pythran/pythonic/include/types/numpy_expr.hpp
@@ -591,28 +591,32 @@ namespace types
     template <class F>
     typename std::enable_if<
         is_numexpr_arg<F>::value &&
-            std::is_same<bool, typename F::dtype>::value,
+            std::is_same<bool, typename F::dtype>::value &&
+            !is_pod_array<F>::value,
         numpy_vexpr<numpy_expr, ndarray<long, pshape<long>>>>::type
     fast(F const &filter) const;
 
     template <class F>
     typename std::enable_if<
         is_numexpr_arg<F>::value &&
-            std::is_same<bool, typename F::dtype>::value,
+            std::is_same<bool, typename F::dtype>::value &&
+            !is_pod_array<F>::value,
         numpy_vexpr<numpy_expr, ndarray<long, pshape<long>>>>::type
     operator[](F const &filter) const;
 
     template <class F> // indexing through an array of indices -- a view
     typename std::enable_if<is_numexpr_arg<F>::value &&
                                 !is_array_index<F>::value &&
-                                !std::is_same<bool, typename F::dtype>::value,
+                                !std::is_same<bool, typename F::dtype>::value &&
+                                !is_pod_array<F>::value,
                             numpy_vexpr<numpy_expr, F>>::type
     operator[](F const &filter) const;
 
     template <class F> // indexing through an array of indices -- a view
     typename std::enable_if<is_numexpr_arg<F>::value &&
                                 !is_array_index<F>::value &&
-                                !std::is_same<bool, typename F::dtype>::value,
+                                !std::is_same<bool, typename F::dtype>::value &&
+                                !is_pod_array<F>::value,
                             numpy_vexpr<numpy_expr, F>>::type
     fast(F const &filter) const;
 

--- a/pythran/pythonic/include/types/numpy_iexpr.hpp
+++ b/pythran/pythonic/include/types/numpy_iexpr.hpp
@@ -226,7 +226,8 @@ namespace types
     template <class F> // indexing through an array of indices -- a view
     typename std::enable_if<is_numexpr_arg<F>::value &&
                                 !is_array_index<F>::value &&
-                                !std::is_same<bool, typename F::dtype>::value,
+                                !std::is_same<bool, typename F::dtype>::value &&
+                                !is_pod_array<F>::value,
                             numpy_vexpr<numpy_iexpr, F>>::type
     operator[](F const &filter) const
     {

--- a/pythran/pythonic/include/types/numpy_texpr.hpp
+++ b/pythran/pythonic/include/types/numpy_texpr.hpp
@@ -95,13 +95,14 @@ namespace types
     template <class F> // indexing through an array of boolean -- a mask
     typename std::enable_if<
         is_numexpr_arg<F>::value &&
-            std::is_same<bool, typename F::dtype>::value && F::value == 1,
+            std::is_same<bool, typename F::dtype>::value && F::value == 1 &&
+            !is_pod_array<F>::value,
         numpy_vexpr<numpy_texpr_2, ndarray<long, pshape<long>>>>::type
     fast(F const &filter) const;
     template <class F> // indexing through an array of boolean -- a mask
     typename std::enable_if<is_numexpr_arg<F>::value &&
                                 std::is_same<bool, typename F::dtype>::value &&
-                                F::value != 1,
+                                F::value != 1 && !is_pod_array<F>::value,
                             numpy_vexpr<ndarray<dtype, pshape<long>>,
                                         ndarray<long, pshape<long>>>>::type
     fast(F const &filter) const;
@@ -109,20 +110,22 @@ namespace types
     template <class F> // indexing through an array of indices -- a view
     typename std::enable_if<
         is_numexpr_arg<F>::value &&
-            !std::is_same<bool, typename F::dtype>::value,
+            !std::is_same<bool, typename F::dtype>::value &&
+            !is_pod_array<F>::value,
         numpy_vexpr<numpy_texpr_2, ndarray<long, pshape<long>>>>::type
     fast(F const &filter) const;
 
     template <class F> // indexing through an array of boolean -- a mask
     typename std::enable_if<
         is_numexpr_arg<F>::value &&
-            std::is_same<bool, typename F::dtype>::value && F::value == 1,
+            std::is_same<bool, typename F::dtype>::value && F::value == 1 &&
+            !is_pod_array<F>::value,
         numpy_vexpr<numpy_texpr_2, ndarray<long, pshape<long>>>>::type
     operator[](F const &filter) const;
     template <class F> // indexing through an array of boolean -- a mask
     typename std::enable_if<is_numexpr_arg<F>::value &&
                                 std::is_same<bool, typename F::dtype>::value &&
-                                F::value != 1,
+                                F::value != 1 && !is_pod_array<F>::value,
                             numpy_vexpr<ndarray<dtype, pshape<long>>,
                                         ndarray<long, pshape<long>>>>::type
     operator[](F const &filter) const;
@@ -130,7 +133,8 @@ namespace types
     template <class F> // indexing through an array of indices -- a view
     typename std::enable_if<
         is_numexpr_arg<F>::value &&
-            !std::is_same<bool, typename F::dtype>::value,
+            !std::is_same<bool, typename F::dtype>::value &&
+            !is_pod_array<F>::value,
         numpy_vexpr<numpy_texpr_2, ndarray<long, pshape<long>>>>::type
     operator[](F const &filter) const;
     auto operator[](long i) const -> decltype(this->fast(i));

--- a/pythran/pythonic/include/types/numpy_vexpr.hpp
+++ b/pythran/pythonic/include/types/numpy_vexpr.hpp
@@ -98,28 +98,32 @@ namespace types
     template <class E> // indexing through an array of boolean -- a mask
     typename std::enable_if<
         is_numexpr_arg<E>::value &&
-            std::is_same<bool, typename E::dtype>::value,
+            std::is_same<bool, typename E::dtype>::value &&
+            !is_pod_array<F>::value,
         numpy_vexpr<numpy_vexpr, ndarray<long, pshape<long>>>>::type
     fast(E const &filter) const;
 
     template <class E> // indexing through an array of boolean -- a mask
     typename std::enable_if<
         is_numexpr_arg<E>::value &&
-            std::is_same<bool, typename E::dtype>::value,
+            std::is_same<bool, typename E::dtype>::value &&
+            !is_pod_array<F>::value,
         numpy_vexpr<numpy_vexpr, ndarray<long, pshape<long>>>>::type
     operator[](E const &filter) const;
 
     template <class E> // indexing through an array of indices -- a view
     typename std::enable_if<is_numexpr_arg<E>::value &&
                                 !is_array_index<E>::value &&
-                                !std::is_same<bool, typename E::dtype>::value,
+                                !std::is_same<bool, typename E::dtype>::value &&
+                                !is_pod_array<F>::value,
                             numpy_vexpr<numpy_vexpr, E>>::type
     operator[](E const &filter) const;
 
     template <class E> // indexing through an array of indices -- a view
     typename std::enable_if<is_numexpr_arg<E>::value &&
                                 !is_array_index<E>::value &&
-                                !std::is_same<bool, typename E::dtype>::value,
+                                !std::is_same<bool, typename E::dtype>::value &&
+                                !is_pod_array<F>::value,
                             numpy_vexpr<numpy_vexpr, E>>::type
     fast(E const &filter) const;
 

--- a/pythran/pythonic/include/utils/numpy_traits.hpp
+++ b/pythran/pythonic/include/utils/numpy_traits.hpp
@@ -59,6 +59,14 @@ namespace types
    * Only used to write concise expression templates
    */
   template <class T>
+  struct is_pod_array {
+    static constexpr bool value = false;
+  };
+  template <class T, size_t N>
+  struct is_pod_array<types::array<T, N>> {
+    static constexpr bool value = true;
+  };
+  template <class T>
   struct is_array {
     static constexpr bool value = false;
   };

--- a/pythran/pythonic/types/ndarray.hpp
+++ b/pythran/pythonic/types/ndarray.hpp
@@ -740,7 +740,8 @@ namespace types
   template <class F> // indexing through an array of boolean -- a mask
   typename std::enable_if<
       is_numexpr_arg<F>::value &&
-          std::is_same<bool, typename F::dtype>::value && F::value == 1,
+          std::is_same<bool, typename F::dtype>::value && F::value == 1 &&
+          !is_pod_array<F>::value,
       numpy_vexpr<ndarray<T, pS>, ndarray<long, pshape<long>>>>::type
   ndarray<T, pS>::fast(F const &filter) const
   {
@@ -759,7 +760,8 @@ namespace types
   template <class F> // indexing through an array of boolean -- a mask
   typename std::enable_if<
       is_numexpr_arg<F>::value &&
-          std::is_same<bool, typename F::dtype>::value && F::value == 1,
+          std::is_same<bool, typename F::dtype>::value && F::value == 1 &&
+          !is_pod_array<F>::value,
       numpy_vexpr<ndarray<T, pS>, ndarray<long, pshape<long>>>>::type
       ndarray<T, pS>::
       operator[](F const &filter) const
@@ -770,7 +772,8 @@ namespace types
   template <class F> // indexing through an array of boolean -- a mask
   typename std::enable_if<
       is_numexpr_arg<F>::value &&
-          std::is_same<bool, typename F::dtype>::value && F::value != 1,
+          std::is_same<bool, typename F::dtype>::value && F::value != 1 &&
+          !is_pod_array<F>::value,
       numpy_vexpr<ndarray<T, pshape<long>>, ndarray<long, pshape<long>>>>::type
   ndarray<T, pS>::fast(F const &filter) const
   {
@@ -782,7 +785,8 @@ namespace types
   template <class F> // indexing through an array of boolean -- a mask
   typename std::enable_if<
       is_numexpr_arg<F>::value &&
-          std::is_same<bool, typename F::dtype>::value && F::value != 1,
+          std::is_same<bool, typename F::dtype>::value && F::value != 1 &&
+          !is_pod_array<F>::value,
       numpy_vexpr<ndarray<T, pshape<long>>, ndarray<long, pshape<long>>>>::type
       ndarray<T, pS>::
       operator[](F const &filter) const
@@ -794,7 +798,8 @@ namespace types
   template <class F> // indexing through an array of indices -- a view
   typename std::enable_if<is_numexpr_arg<F>::value &&
                               !is_array_index<F>::value &&
-                              !std::is_same<bool, typename F::dtype>::value,
+                              !std::is_same<bool, typename F::dtype>::value &&
+                              !is_pod_array<F>::value,
                           numpy_vexpr<ndarray<T, pS>, F>>::type ndarray<T, pS>::
   operator[](F const &filter) const
   {
@@ -805,7 +810,8 @@ namespace types
   template <class F> // indexing through an array of indices -- a view
   typename std::enable_if<is_numexpr_arg<F>::value &&
                               !is_array_index<F>::value &&
-                              !std::is_same<bool, typename F::dtype>::value,
+                              !std::is_same<bool, typename F::dtype>::value &&
+                              !is_pod_array<F>::value,
                           numpy_vexpr<ndarray<T, pS>, F>>::type
   ndarray<T, pS>::fast(F const &filter) const
   {

--- a/pythran/pythonic/types/numpy_expr.hpp
+++ b/pythran/pythonic/types/numpy_expr.hpp
@@ -293,7 +293,9 @@ namespace types
   template <class Op, class... Args>
   template <class F>
   typename std::enable_if<
-      is_numexpr_arg<F>::value && std::is_same<bool, typename F::dtype>::value,
+      is_numexpr_arg<F>::value &&
+          std::is_same<bool, typename F::dtype>::value &&
+          !is_pod_array<F>::value,
       numpy_vexpr<numpy_expr<Op, Args...>, ndarray<long, pshape<long>>>>::type
   numpy_expr<Op, Args...>::fast(F const &filter) const
   {
@@ -312,7 +314,9 @@ namespace types
   template <class Op, class... Args>
   template <class F>
   typename std::enable_if<
-      is_numexpr_arg<F>::value && std::is_same<bool, typename F::dtype>::value,
+      is_numexpr_arg<F>::value &&
+          std::is_same<bool, typename F::dtype>::value &&
+          !is_pod_array<F>::value,
       numpy_vexpr<numpy_expr<Op, Args...>, ndarray<long, pshape<long>>>>::type
       numpy_expr<Op, Args...>::
       operator[](F const &filter) const
@@ -323,7 +327,8 @@ namespace types
   template <class F> // indexing through an array of indices -- a view
   typename std::enable_if<is_numexpr_arg<F>::value &&
                               !is_array_index<F>::value &&
-                              !std::is_same<bool, typename F::dtype>::value,
+                              !std::is_same<bool, typename F::dtype>::value &&
+                              !is_pod_array<F>::value,
                           numpy_vexpr<numpy_expr<Op, Args...>, F>>::type
       numpy_expr<Op, Args...>::
       operator[](F const &filter) const
@@ -335,7 +340,8 @@ namespace types
   template <class F> // indexing through an array of indices -- a view
   typename std::enable_if<is_numexpr_arg<F>::value &&
                               !is_array_index<F>::value &&
-                              !std::is_same<bool, typename F::dtype>::value,
+                              !std::is_same<bool, typename F::dtype>::value &&
+                              !is_pod_array<F>::value,
                           numpy_vexpr<numpy_expr<Op, Args...>, F>>::type
   numpy_expr<Op, Args...>::fast(F const &filter) const
   {

--- a/pythran/pythonic/types/numpy_texpr.hpp
+++ b/pythran/pythonic/types/numpy_texpr.hpp
@@ -158,7 +158,8 @@ namespace types
   template <class F> // indexing through an array of boolean -- a mask
   typename std::enable_if<
       is_numexpr_arg<F>::value &&
-          std::is_same<bool, typename F::dtype>::value && F::value == 1,
+          std::is_same<bool, typename F::dtype>::value && F::value == 1 &&
+          !is_pod_array<F>::value,
       numpy_vexpr<numpy_texpr_2<E>, ndarray<long, pshape<long>>>>::type
   numpy_texpr_2<E>::fast(F const &filter) const
   {
@@ -176,7 +177,8 @@ namespace types
   template <class F> // indexing through an array of boolean -- a mask
   typename std::enable_if<
       is_numexpr_arg<F>::value &&
-          std::is_same<bool, typename F::dtype>::value && F::value != 1,
+          std::is_same<bool, typename F::dtype>::value && F::value != 1 &&
+          !is_pod_array<F>::value,
       numpy_vexpr<ndarray<typename numpy_texpr_2<E>::dtype, pshape<long>>,
                   ndarray<long, pshape<long>>>>::type
   numpy_texpr_2<E>::fast(F const &filter) const
@@ -189,7 +191,8 @@ namespace types
   template <class F> // indexing through an array of boolean -- a mask
   typename std::enable_if<
       is_numexpr_arg<F>::value &&
-          std::is_same<bool, typename F::dtype>::value && F::value == 1,
+          std::is_same<bool, typename F::dtype>::value && F::value == 1 &&
+          !is_pod_array<F>::value,
       numpy_vexpr<numpy_texpr_2<E>, ndarray<long, pshape<long>>>>::type
       numpy_texpr_2<E>::
       operator[](F const &filter) const
@@ -201,7 +204,8 @@ namespace types
   template <class F> // indexing through an array of boolean -- a mask
   typename std::enable_if<
       is_numexpr_arg<F>::value &&
-          std::is_same<bool, typename F::dtype>::value && F::value != 1,
+          std::is_same<bool, typename F::dtype>::value && F::value != 1 &&
+          !is_pod_array<F>::value,
       numpy_vexpr<ndarray<typename numpy_texpr_2<E>::dtype, pshape<long>>,
                   ndarray<long, pshape<long>>>>::type numpy_texpr_2<E>::
   operator[](F const &filter) const
@@ -212,7 +216,9 @@ namespace types
   template <class E>
   template <class F> // indexing through an array of indices -- a view
   typename std::enable_if<
-      is_numexpr_arg<F>::value && !std::is_same<bool, typename F::dtype>::value,
+      is_numexpr_arg<F>::value &&
+          !std::is_same<bool, typename F::dtype>::value &&
+          !is_pod_array<F>::value,
       numpy_vexpr<numpy_texpr_2<E>, ndarray<long, pshape<long>>>>::type
       numpy_texpr_2<E>::
       operator[](F const &filter) const
@@ -226,12 +232,14 @@ namespace types
   template <class E>
   template <class F> // indexing through an array of indices -- a view
   typename std::enable_if<
-      is_numexpr_arg<F>::value && !std::is_same<bool, typename F::dtype>::value,
+      is_numexpr_arg<F>::value &&
+          !std::is_same<bool, typename F::dtype>::value &&
+          !is_pod_array<F>::value,
       numpy_vexpr<numpy_texpr_2<E>, ndarray<long, pshape<long>>>>::type
   numpy_texpr_2<E>::fast(F const &filter) const
   {
     static_assert(F::value == 1,
-                  "advanced indexing only supporint with 1D index");
+                  "advanced indexing only supported with 1D index");
     return {*this, filter};
   }
 

--- a/pythran/pythonic/types/numpy_vexpr.hpp
+++ b/pythran/pythonic/types/numpy_vexpr.hpp
@@ -89,7 +89,9 @@ namespace types
   template <class T, class F>
   template <class E> // indexing through an array of boolean -- a mask
   typename std::enable_if<
-      is_numexpr_arg<E>::value && std::is_same<bool, typename E::dtype>::value,
+      is_numexpr_arg<E>::value &&
+          std::is_same<bool, typename E::dtype>::value &&
+          !is_pod_array<F>::value,
       numpy_vexpr<numpy_vexpr<T, F>, ndarray<long, pshape<long>>>>::type
   numpy_vexpr<T, F>::fast(E const &filter) const
   {
@@ -108,7 +110,9 @@ namespace types
   template <class T, class F>
   template <class E> // indexing through an array of boolean -- a mask
   typename std::enable_if<
-      is_numexpr_arg<E>::value && std::is_same<bool, typename E::dtype>::value,
+      is_numexpr_arg<E>::value &&
+          std::is_same<bool, typename E::dtype>::value &&
+          !is_pod_array<F>::value,
       numpy_vexpr<numpy_vexpr<T, F>, ndarray<long, pshape<long>>>>::type
       numpy_vexpr<T, F>::
       operator[](E const &filter) const
@@ -120,7 +124,8 @@ namespace types
   template <class E> // indexing through an array of indices -- a view
   typename std::enable_if<is_numexpr_arg<E>::value &&
                               !is_array_index<E>::value &&
-                              !std::is_same<bool, typename E::dtype>::value,
+                              !std::is_same<bool, typename E::dtype>::value &&
+                              !is_pod_array<F>::value,
                           numpy_vexpr<numpy_vexpr<T, F>, E>>::type
       numpy_vexpr<T, F>::
       operator[](E const &filter) const
@@ -132,7 +137,8 @@ namespace types
   template <class E> // indexing through an array of indices -- a view
   typename std::enable_if<is_numexpr_arg<E>::value &&
                               !is_array_index<E>::value &&
-                              !std::is_same<bool, typename E::dtype>::value,
+                              !std::is_same<bool, typename E::dtype>::value &&
+                              !is_pod_array<F>::value,
                           numpy_vexpr<numpy_vexpr<T, F>, E>>::type
   numpy_vexpr<T, F>::fast(E const &filter) const
   {

--- a/pythran/tests/test_ndarray.py
+++ b/pythran/tests/test_ndarray.py
@@ -911,3 +911,12 @@ def assign_ndarray(t):
                       numpy.ndarray((2,2), numpy.float32),
                       numpy.ndarray((2,2), numpy.bool),
                       vexpr_of_texpr=[NDArray[numpy.float32,:,:], NDArray[numpy.bool,:,:]])
+
+    def test_indexing_through_int8(self):
+        code = '''
+            def indexing_through_int8(x):
+                return x[x[0,0],x[0,1]]'''
+        self.run_test(code,
+                      numpy.arange(10, dtype=numpy.uint8).reshape(5,2),
+                      indexing_through_int8=[NDArray[numpy.uint8,:,:]])
+

